### PR TITLE
fix(sdk): fix behavior of Run in disabled mode

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------

Ensure that all public attributes and methods of the Run object are available in disabled mode.

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

### CODE REVIEW

File: wandb/sdk/wandb_init.py [hunk 1]
```diff
@@ -522,17 +525,60 @@ def _log_setup(self, settings: Settings) -> None:
         logger.info(f"Logging internal logs to {settings.log_internal}")
 
     def _make_run_disabled(self) -> Run:
+        """Create a disabled run object.
```
Comment (timoffex) [inline wandb/sdk/wandb_init.py:528]:
```suggestion
        """Returns a Run-like object where all methods are no-ops.
```

File: wandb/sdk/wandb_init.py [hunk 2]
```diff
@@ -522,17 +525,60 @@ def _log_setup(self, settings: Settings) -> None:
...
+
+        This enables the user to call wandb.init(mode="disabled") and still have a Run object.
+        We ensure that the attributes and methods of the run object are available, but do nothing.
```
Comment (timoffex) [inline wandb/sdk/wandb_init.py:531]: Should probably say more here. How does this help users?

Maybe something like

> Although it's inadvisable, this allows users to pass a no-op Run object to APIs that expect and call methods on a Run. Without this, users encounter attribute errors.

Even better if we have any specific examples.

File: wandb/sdk/wandb_init.py [hunk 3]
```diff
@@ -174,7 +175,9 @@ def setup(self, kwargs: Any) -> None:  # noqa: C901
         # we add this logic to be backward compatible with the old behavior of disable
         # where it would disable the service if the mode was set to disabled
         mode = kwargs.get("mode")
-        settings_mode = (kwargs.get("settings") or {}).get("mode")
+        settings_mode = (kwargs.get("settings") or {}).get("mode") or os.environ.get(
+            "WANDB_MODE"
+        )
```
Comment (timoffex) [inline wandb/sdk/wandb_init.py:180]: Isn't this duplicating logic that exists in `Settings` that already reads the environment?

**Review by timoffex — APPROVED** [review]

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -2158,12 +2159,13 @@ def _finish(
     @_run_decorator._attach
     def join(self, exit_code: Optional[int] = None) -> None:
         """Deprecated alias for `finish()` - use finish instead."""
-        deprecate.deprecate(
-            field_name=deprecate.Deprecated.run__join,
-            warning_message=(
-                "wandb.run.join() is deprecated, please use wandb.run.finish()."
-            ),
-        )
+        if hasattr(self, "_telemetry_obj"):
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply wandb/sdk/wandb_run.py:2162]: oops, forgot to delete this, thanks!

[the remark this answers is not in the document]
Comment (dmitryduev) [reply wandb/sdk/wandb_run.py:2162]: would love to delete join() altogether, but i just looked at telemetry, and oh boy do people still use it :(

File: wandb/sdk/wandb_init.py (hunk 3 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_init.py:180]: we don't parse env var until after this point, so what was happening - we were still spinning up core if disabled via an env var...

**Pull request merged**